### PR TITLE
Add `.slice_by_idx()` methods to gammapy maps

### PR DESCRIPTION
**Pull request by adonath**
This PR adds `.slice_by_idx()` methods to the `MapGeom` and `Map` classes. I'm opening this PR early to get some feedback on the implementation and API, before I continue with  `.slice_by_coord()`. Does the `drop_axes` keyword makes sense to you (this was introduced by @woodmd in `Map.to_slice()`) or shall we rather implement a `Map.drop_axis()` method? I introduced a hidden `_copy_init_kwargs` property to `WcsGeom` and `HpxGeom` to easily create a new `WcsGeom` and `HpxGeom` object with the same specification. Or shall we rather introduce a `WcsGeom.copy()` method?

### CODE REVIEW

**Comment by cdeil** [comment]
Another comment: can you please add a few more explicit tests, e.g. of a 4D map with `event_id = [7, 9, 42]` and `energy = [3, 5, 7, 9]`, maybe even with a single spatial bin?

One thing to assert on is the output shape when slicing, but also checking that the sliced map data is a view would be important, so that we don't accidentally make copies. I've never done this, but looking at https://stackoverflow.com/a/11524746/498873 I think this might do it:
```
m2 = m.slice(...)
assert m2.data.base is m.data
```

**Comment by cdeil** [comment]
If you don't think a `MapSlice` is a good idea, because it's only used in one place, and doesn't have enough code to warrant splitting it out into a separate class, fine.

My main suggestion is to support this:
```
map.slice_by_idx({'energy': slice(3, 7), 'time': 42)
```
and to drop the `drop_axes` from the slice method on map and geom.

**Comment by cdeil** [comment]
@adonath - I think you added a new "Indexing and Slicing" section in the docs, but there was already an empty one further down on that docs page:
http://docs.gammapy.org/dev/maps/index.html#slicing-methods
Not sure whether it fits better further up or down, but suggest to remove the empty one.

Otherwise, 👍 , this is great!

File: docs/maps/index.rst [hunk 1]
```diff
@@ -125,6 +125,43 @@ Fermi-LAT PSF:
    m_hpx = Map.create(binsz=binsz, map_type='hpx', skydir=position, width=10.0,
                           axes=[energy_axis])
 
+Indexing and Slicing
+--------------------
```
Comment (cdeil) [inline docs/maps/index.rst]: I think there needs to be an empty line after a section heading in RST for this to render correctly? (not 100% sure, didn't try)

File: gammapy/maps/base.py [hunk 2]
```diff
@@ -499,6 +499,32 @@ def upsample(self, factor, order=0, preserve_counts=True):
         """
         pass
 
+    def slice_by_idx(self, slices, copy=False):
+        """Slice sub map from map object.
+
```
Comment (cdeil) [inline gammapy/maps/base.py]: Add reference to the high-level docs section with the examples?

Comment (adonath) [reply gammapy/maps/base.py]: Done

File: docs/maps/index.rst (hunk 1 above)
Comment (adonath) [reply docs/maps/index.rst]: Done

**Comment by adonath** [comment]
@cdeil Thanks for your comments! For now I decided against the `MapSlice` object, but we should keep the idea in the back of our head, once we introduce the `.slice_by_coord()` method. But still the argument remains, that it will be only used in one place...

I'll go ahead an merge, the failing test is unrelated.

**Comment by cdeil** [comment]
🍴 Hooray! 🍴

**Pull request merged**